### ISSUE TITLE
Output of tests to stdout

### ISSUE THREAD

**Issue by Manuzor** [fastbuild/fastbuild#106]
The `Test()` function has the required parameter `.TestOutput` which, as far as I can tell, can only be set to a filepath on disk. I would like to see other options here, such as redirecting it directly to stdout.

I use FASTBuild from within Sublime Text which captures stdout and presents that live to the user. Seeing the output of my unit tests from within the editor, especially when they fail, would be very useful.

Or is this possible already and I missed something in the docs? Maybe with a magic string that couldn't be a path, such as `.TestOutput = ':stdout'`?

**Comment by ffulin** [comment]
Hello

The behaviour of Test() is as follows:
 - A single line is emitted to stdout when the test starts - e.g. "1> Running Test c:\path\file"
 - The output of your test (stdout and stderr) is captured while the test is running, and is written to the file specified by .TestOutput (whether the test passed or fails)
 - Additionally, if the test fails, the captured output will also be written to the stdout

The design goal here is that FASTBuild doesn't emit any extraneous output unless there is a warning or failure.  I think this works well for tests, where you typically only care about the output when the test fails.  Your test can be quite spammy without overwhelming the output, since that output is only seen on failure.

Having said that, it would be trivial to add a .TestAlwaysWriteToStdOut option if it was desirable to see the output in all cases.  With the current implementation it would print at the end of the test.

**Comment by Manuzor** [comment]
Hi @ffulin,

thanks for your feedback! Silly me, I didn't think to try out what happens when my tests failed. :smile: The behavior you described sounds really useful and thoughtful!

You mentioned that FASTBuild will only emit the test output in case of a warning or failure but how do you determine that? Are you parsing the test output or are you relying on the exit code of the process?

You're writing the documentation all by yourself, right? If so, did you consider setting up a repo for it and accepting pull requests for it? I'd be glad to contribute based on our discussion here.

In any case, I still think it would be nice to have the option to always show the output of a test. You said it's trivial to add this option _with the current implementation_ but that means that it will wait for the process to terminate and _then_ print the results, right? I think in case of printing to stdout it would be useful to see the output live, i.e. directly wiring the test output to stdout of FASTBuild.

**Comment by ffulin** [comment]
Hello again :)

> Are you parsing the test output or are you relying on the exit code of the process?

For tests, the return code is checked (a standard non-zero indicating failure).  There is no concept of warnings for tests.  When compiling, the return code is checked and the output is parsed to check for warnings.

> If so, did you consider setting up a repo for it and accepting pull requests for it?

The documentation shown on the website is part of the source code, so I can happily accept pull requests for that like anything else.  The root of the documentation is here: https://github.com/fastbuild/fastbuild/tree/dev/Code/Tools/FBuild/Documentation
The level of documentation varies from function to function, and I'm working through updating them as I go. Test() is one I have not gotten to yet. Help expanding the documentation would be appreciated. An example of a function with more complete documentation is Alias (http://fastbuild.org/docs/functions/alias.html). You could use that as a style/layout guide.

> I think in case of printing to stdout it would be useful to see the output live, 

There are two things to consider here.

**a) Modifying the process spawner.** This would require changes to the core process spawning code, including changes on all platforms.  Non-trivial, but doable.

**b) Handling parallelism.**  Multiple tests could run in parallel, in which case outputting the test output as it's captured would mix the output from several tests, making it very difficult to understand.  The way FASTBuild is designed, outputting at the end of a task if necessary and doing it all in one chunk.  This is really the main problem.  There isn't a good solution for this that I can see.

**Comment by Manuzor** [comment]
(Sorry for the late reply)

> b) Handling parallelism.

You are right, of course. When running tests in parallel, each line would have to be prefixed or uniquely highlighting (e.g. colors) to distinguish the different test runs. Visual Studio, for example, does this by assigning numbers to build jobs and prefixing output messages with that number.

In the end I don't think it is worth the effort, especially since you can always run the executable on its own to achieve the result (just like you mentioned). Maybe this could be noted in the documentation.

Speaking of which, I've started working on the documentation and will submit a PR soon.

**Comment by Manuzor** [comment]
Oh, I forgot to mention something. I've been using FASTBuild actively for a while now, using the Test() function a lot and I'm quite happy with its behavior. I like the design of keeping the output clean as much as possible. At first I was confused, thinking FASTBuild always swallows the output no matter what because I only tested it on successful test runs. That's what motivated me to write this issue in the first place.

You mentioned it would be trivial to implement something like a `.TestAlwaysWriteToStdOut` option. Do you think it should be added, in case someone else needs this kind of option?

**Comment by ffulin** [comment]
> You mentioned it would be trivial to implement something like a .TestAlwaysWriteToStdOut option. Do you think it should be added, in case someone else needs this kind of option?

I have a long backlog of things to implement/fix/improve that I would consider better targets to work on, so even though this would be pretty easy, I probably wouldn't get around to it for quite some time. I would happily accept a pull request if this feature was important enough to someone that they would implement it :)

**Comment by ffulin** [comment]
We can keep this issue open to track this enhancement.

**Comment by ffulin** [comment]
So it only took 5 minutes so I have added a new .TestAlwaysShowOutput option. I've integrated it into the 'dev' branch. It will be in the next release (v0.92).  I'll leave this issue open until that release is available.

**Comment by Manuzor** [comment]
Thanks for adding this, even though your schedule is tight! I was going to attempt to add it, but that would probably have taken more than 5 minutes. :smile:

**Comment by ffulin** [comment]
v0.92 has now been released and contains the new option. Thanks again for your help!
